routing/payment_lifecycle.go

```diff
@@ -686,6 +688,38 @@ func (p *paymentLifecycle) sendAttempt(
 		CustomRecords: lnwire.CustomRecords(p.firstHopTLVs),
 	}
 
+	// If a hook exists that may affect our outgoing message, we call it now
+	// and apply its side effects to the UpdateAddHTLC message.
+	err := fn.MapOptionZ(
+		p.router.cfg.TrafficShaper,
+		func(ts TlvTrafficShaper) error {
+			newAmt, newData, err := ts.ProduceHtlcExtraData(
+				rt.TotalAmount, htlcAdd.CustomRecords,
+			)
+			if err != nil {
+				return err
+			}
+
+			customRecords, err := lnwire.ParseCustomRecords(newData)
+			if err != nil {
+				return err
+			}
+
+			log.Debugf("TLV traffic shaper returned custom "+
+				"records %v and amount %d msat for HTLC",
+				spew.Sdump(customRecords), newAmt)
+
+			htlcAdd.CustomRecords = customRecords
+			htlcAdd.Amount = newAmt
```

Comment:
once again I *strongly* discourage mutating outer scope values in these closures. If you insist on doing so, don't use the Option type, since the general expectation about the usage of MapOptionZ is to not have side effects.